javatests/arcs/android/storage/service/StorageServiceNgImplTest.kt

```diff
@@ -0,0 +1,176 @@
+import kotlinx.coroutines.test.TestCoroutineScope
+import kotlinx.coroutines.test.runBlockingTest
+import org.junit.After
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+
+@RunWith(AndroidJUnit4::class)
+class StorageServiceNgImplTest {
+
+  private val driverFactory = FixedDriverFactory(MockDriverProvider())
+  private val onProxyMessageCallback: suspend (StorageKey, UntypedProxyMessage) -> Unit =
+    { _: StorageKey, _: UntypedProxyMessage -> }
+  private val messageCallback = FakeMessageCallback()
+  private val channelCallback = FakeStorageChannelCallback()
+
+  private val scope = TestCoroutineScope()
+  private lateinit var storageService: StorageServiceNgImpl
+
+  @Before
+  fun setUp() {
+    StorageKeyParser.addParser(RamDiskStorageKey)
+    storageService = StorageServiceNgImpl(
+      scope,
+      driverFactory,
+      ::testWriteBackProvider,
+      null,
+      onProxyMessageCallback
+    )
+  }
+
+  @After
+  fun tearDown() {
+    scope.cleanupTestCoroutines()
+  }
+
+  @Test
+  fun openStorageChannel_invokesOnCreateCallback_andCreatesChannel() = scope.runBlockingTest {
+    val encodedStoreOptions = createEncodedStoreOptions(
+      RamDiskStorageKey("myCount"),
+      CountType(),
+      ParcelableCrdtType.Count
+    )
+
+    storageService.openStorageChannel(encodedStoreOptions, channelCallback, messageCallback)
+
+    // `waitForOnCreate` ensures the `onCreate` callback was invoked.
+    val channel = channelCallback.waitForOnCreate()
+
+    assertThat(channel).isInstanceOf(StorageChannelImpl::class.java)
+  }
+
+  @Test
+  fun openingTwoChannels_forSameStore_resultsInTwoDistinctChannels() = scope.runBlockingTest {
+    val encodedStoreOptions = createEncodedStoreOptions(
+      RamDiskStorageKey("myCount"),
+      CountType(),
+      ParcelableCrdtType.Count
+    )
+
+    val channelCallback2 = FakeStorageChannelCallback()
+    storageService.openStorageChannel(encodedStoreOptions, channelCallback, messageCallback)
+    storageService.openStorageChannel(encodedStoreOptions, channelCallback2, messageCallback)
+
+    val channel1 = channelCallback.waitForOnCreate()
+    val channel2 = channelCallback2.waitForOnCreate()
+    assertThat(channel2).isNotSameInstanceAs(channel1)
+  }
+
+  @Test
+  fun openStorageChannel_createsAStore() = scope.runBlockingTest {
+    val storageKey = RamDiskStorageKey("store1")
+    val encodedStoreOptions = createEncodedStoreOptions(
+      storageKey,
+      CountType(),
+      ParcelableCrdtType.Count
+    )
+
+    storageService.openStorageChannel(encodedStoreOptions, channelCallback, messageCallback)
+    val channel = channelCallback.waitForOnCreate() as StorageChannelImpl
+
+    assertThat(storageService.activeStorageKeys()).containsExactly(storageKey)
+    assertThat(channel.storageKey).isEqualTo(storageKey)
+  }
+
+  @Test
+  fun openingTwoChannels_forDifferentStores_createsTwoStores() = scope.runBlockingTest {
```

Comment:
Using an Arrange, Act, Assert structure, with some clearer grouping of the pairs of variables, in this test would look something like this:

```
    // Arrange.
    // Define the store options for two different stores.
    val storageKey1 = RamDiskStorageKey("store1")
    val storageKey2 = RamDiskStorageKey("store2")
    val encodedStoreOptions1 = createEncodedStoreOptions(
      storageKey1,
      CountType(),
      ParcelableCrdtType.Count
    )
    val encodedStoreOptions2 = createEncodedStoreOptions(
      storageKey2,
      CountType(),
      ParcelableCrdtType.Count
    )
    val channelCallback1 = FakeStorageChannelCallback()
    val channelCallback2 = FakeStorageChannelCallback()

    // Act.
    storageService.openStorageChannel(encodedStoreOptions1, channelCallback1, messageCallback)
    storageService.openStorageChannel(encodedStoreOptions2, channelCallback2, messageCallback)

    // Assert.
    val channel1 = channelCallback1.waitForOnCreate() as StorageChannelImpl
    val channel2 = channelCallback2.waitForOnCreate() as StorageChannelImpl
    // Check two stores have been created.
    assertThat(storageService.activeStorageKeys())
      .containsExactlyElementsIn(listOf(storageKey1, storageKey2))
    // Check the channels are communicating with the expected stores.
    assertThat(channel1.storageKey).isEqualTo(storageKey1)
    assertThat(channel2.storageKey).isEqualTo(storageKey2)
```

All of the comments (both the descriptive ones you have and the Arrange/Act/Assert ones) are optional if you think it's clear without them. If you stick to the three block structure (three blocks with a single blank line separating them) you almost certainly don't need the latter comments.

Reply:
Yes I agree this is elegant. I left out the Arrange/ActAssert comments but kept the descriptive ones. I think the code is organised in a way that makes the three separate blocks clear.